tests/unit_tests/test_solver.cpp

```diff
@@ -1,135 +1,172 @@
-  EXPECT_NEAR(solution(2), exactSolution(2), std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(0), exact_solution(0), std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(1), exact_solution(1), std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(2), exact_solution(2), std::numeric_limits<double>::epsilon());
 }
 
-TEST(LinearSolverTest, solve_2x2_matrix)
-{
-  auto matrix = create_matrix({{1., 2.}, {3., 4.}});
-  auto solution = create_vector({17., 39.});
-  auto exactSolution = create_vector({5., 6.});
+TEST(LinearSolverTest, Solve2x2Matrix) {
+    auto matrix = create_matrix({{1., 2.}, {3., 4.}});
+    auto solution = create_vector({17., 39.});
+    auto exact_solution = create_vector({5., 6.});
 
-  auto info = openturbine::rigid_pendulum::solve_linear_system(matrix, solution);
+    auto info = openturbine::rigid_pendulum::solve_linear_system(matrix, solution);
 
-  ASSERT_EQ(info, 0);
+    ASSERT_EQ(info, 0);
 
-  EXPECT_NEAR(solution(0), exactSolution(0), 10*std::numeric_limits<double>::epsilon());
-  EXPECT_NEAR(solution(1), exactSolution(1), 10*std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(0), exact_solution(0), 10 * std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(1), exact_solution(1), 10 * std::numeric_limits<double>::epsilon());
 }
 
-TEST(LinearSolverTest, solve_3x3_matrix)
-{
-  auto matrix = create_matrix({{2., 6., 3.}, {4., -1., 3.}, {1., 3., 2.}});
-  auto solution = create_vector({23., 11., 13.});
-  auto exactSolution = create_vector({1., 2., 3.});
+TEST(LinearSolverTest, Solve3x3Matrix) {
+    auto matrix = create_matrix({{2., 6., 3.}, {4., -1., 3.}, {1., 3., 2.}});
+    auto solution = create_vector({23., 11., 13.});
+    auto exact_solution = create_vector({1., 2., 3.});
 
-  auto info = openturbine::rigid_pendulum::solve_linear_system(matrix, solution);
+    auto info = openturbine::rigid_pendulum::solve_linear_system(matrix, solution);
 
-  ASSERT_EQ(info, 0);
+    ASSERT_EQ(info, 0);
+
+    EXPECT_NEAR(solution(0), exact_solution(0), 10 * std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(1), exact_solution(1), 10 * std::numeric_limits<double>::epsilon());
+    EXPECT_NEAR(solution(2), exact_solution(2), 10 * std::numeric_limits<double>::epsilon());
+}
 
-  EXPECT_NEAR(solution(0), exactSolution(0), 10*std::numeric_limits<double>::epsilon());
-  EXPECT_NEAR(solution(1), exactSolution(1), 10*std::numeric_limits<double>::epsilon());
-  EXPECT_NEAR(solution(2), exactSolution(2), 10*std::numeric_limits<double>::epsilon());
+TEST(LinearSolverTest, CheckMatrixShape) {
+    // Try to solve a non-square system
+    auto matrix_2x3 = create_matrix({{1., 2., 3.}, {4., 5., 6.}});
+    auto solution_2x1 = create_vector({1., 1.});
+
+    EXPECT_THROW(openturbine::rigid_pendulum::solve_linear_system(matrix_2x3, solution_2x1),
+                 std::invalid_argument);
+
+    // Try to solve another non-square system
```

Comment:
I'd make this two separate tests

Reply:
I am not well-versed in the standard practices of TDD, so pardon my ignorance here. Do we always need to write separate tests every time we introduce new inputs? To be specific, for the tests here they are checking if the matrix is square or not - is it not recommended to assert several matrices even though you reach the solution in production code in the first pass?

Reply:
In general, I like to make every set of inputs its own test. This keeps our tests really simple and easy to understand. Also, a failure points you right to the failure points and there are very few moving parts to complicate things.  

You should definitely add as many tests as reasonable, even once they all pass, but keeping them separate aids in writing them now and future reading and debugging.  Similar to how you want your functions to have a single responsibility.